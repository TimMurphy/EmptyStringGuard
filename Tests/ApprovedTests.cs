#if(DEBUG)
using Microsoft.Build.Utilities;
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using ApprovalTests;
using ApprovalTests.Reporters;
using NUnit.Framework;

[Ignore("ToolLocationHelper.GetPathToDotNetFrameworkSdk(TargetDotNetFrameworkVersion.Version40) returns null on my machine")]
[TestFixture]
[UseReporter(typeof(DiffReporter), typeof(ClipboardReporter))]
public class ApprovedTests
{
    [Test]
    public void ClassWithBadAttributes()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "ClassWithBadAttributes"));
    }

    [Test]
    public void ClassWithPrivateMethod()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "ClassWithPrivateMethod"));
    }

    [Test]
    public void InterfaceBadAttributes()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "InterfaceBadAttributes"));
    }

    [Test]
    public void SimpleClass()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "SimpleClass"));
    }

    [Test]
    public void SkipIXamlMetadataProvider()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "XamlMetadataProvider"));
    }

    [Test]
    public void SpecialClass()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "SpecialClass"));
    }

    private static string Decompile(string assemblyPath, string identifier = "")
    {
        var exePath = GetPathToILDasm();

        if (!string.IsNullOrEmpty(identifier))
            identifier = "/item:" + identifier;

        using (var process = Process.Start(new ProcessStartInfo(exePath, $"\"{assemblyPath}\" /nobar /text {identifier}")
        {
            RedirectStandardOutput = true,
            UseShellExecute = false,
            CreateNoWindow = true
        }))
        {

            var projectFolder = Path.GetFullPath(Path.GetDirectoryName(TestContext.CurrentContext.TestDirectory) + @"\..\..").Replace(@"\", @"\\");
            projectFolder = $@"{projectFolder[0]}{projectFolder.Substring(1)}\\";

            process.WaitForExit(10000);

            var decompile = process.StandardOutput.ReadToEnd()
                .Replace(projectFolder, "");

            return decompile.Substring(decompile.IndexOf("// warning :"));
        }
    }

    private static string GetPathToILDasm()
    {
        var path = Path.Combine(ToolLocationHelper.GetPathToDotNetFrameworkSdk(TargetDotNetFrameworkVersion.Version40), @"bin\NETFX 4.0 Tools\ildasm.exe");
        if (!File.Exists(path))
            path = path.Replace("v7.0", "v8.0");
        if (!File.Exists(path))
            Assert.Ignore("ILDasm could not be found");
        return path;
    }
}

#endif