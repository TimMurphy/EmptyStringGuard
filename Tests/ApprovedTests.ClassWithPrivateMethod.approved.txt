// warning : THIS IS A PARTIAL DISASSEMBLY, NOT SUITABLE FOR RE-ASSEMBLING



.class public auto ansi beforefieldinit ClassWithPrivateMethod
       extends [mscorlib]System.Object
{
  .field private string '<SomeProperty>k__BackingField'
  .custom instance void [mscorlib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = ( 01 00 00 00 ) 
  .method public hidebysig instance void 
          PublicWrapperOfPrivateMethod() cil managed
  {
    // Code size       12 (0xc)
    .maxstack  2
    IL_0000:  ldarg.0
    IL_0001:  ldsfld     string [mscorlib]System.String::Empty
    IL_0006:  call       instance void ClassWithPrivateMethod::SomePrivateMethod(string)
    IL_000b:  ret
  } // end of method ClassWithPrivateMethod::PublicWrapperOfPrivateMethod

  .method private hidebysig instance void 
          SomePrivateMethod(string x) cil managed
  {
    // Code size       30 (0x1e)
    .maxstack  2
    IL_0000:  ldarg.1
    IL_0001:  ldsfld     string [mscorlib]System.String::Empty
    IL_0006:  call       bool [mscorlib]System.String::op_Equality(string,
                                                                   string)
    IL_000b:  brfalse.s  IL_001d

    IL_000d:  ldstr      "[EmptyStringGuard] x is an empty string."
    IL_0012:  ldstr      "x"
    IL_0017:  newobj     instance void [mscorlib]System.ArgumentException::.ctor(string,
                                                                                 string)
    IL_001c:  throw

    IL_001d:  ret
  } // end of method ClassWithPrivateMethod::SomePrivateMethod

  .method public hidebysig specialname instance string 
          get_SomeProperty() cil managed
  {
    .custom instance void [mscorlib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = ( 01 00 00 00 ) 
    // Code size       7 (0x7)
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  ldfld      string ClassWithPrivateMethod::'<SomeProperty>k__BackingField'
    IL_0006:  ret
  } // end of method ClassWithPrivateMethod::get_SomeProperty

  .method public hidebysig specialname instance void 
          set_SomeProperty(string 'value') cil managed
  {
    .custom instance void [mscorlib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = ( 01 00 00 00 ) 
    // Code size       8 (0x8)
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  ldarg.1
    IL_0002:  stfld      string ClassWithPrivateMethod::'<SomeProperty>k__BackingField'
    IL_0007:  ret
  } // end of method ClassWithPrivateMethod::set_SomeProperty

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method ClassWithPrivateMethod::.ctor

  .property instance string SomeProperty()
  {
    .get instance string ClassWithPrivateMethod::get_SomeProperty()
    .set instance void ClassWithPrivateMethod::set_SomeProperty(string)
  } // end of property ClassWithPrivateMethod::SomeProperty
} // end of class ClassWithPrivateMethod

// *********** DISASSEMBLY COMPLETE ***********************
